crates/wasi-common/src/hostcalls_impl/fs.rs

```diff
@@ -559,7 +559,8 @@ pub(crate) unsafe fn path_open(
             | wasi::__WASI_RIGHT_FD_WRITE
             | wasi::__WASI_RIGHT_FD_ALLOCATE
             | wasi::__WASI_RIGHT_FD_FILESTAT_SET_SIZE)
-        != 0;
+        != 0
+        || oflags & wasi::__WASI_O_TRUNC != 0;
```

Comment:
If we don't have the right to do `fd_write`, I guess we should be able to have the possibility of truncating the files either. This sounds like a potential DOS vulnerability: we don't have the right to write, but we open all the files that we can and truncate them.

Btw. we probably should add a check for `O_TRUNC` in `path_open` for Linux because using `O_TRUNC` with `O_RDONLY` is undefined. I'll do it. I have no idea how will  `FILE_GENERIC_READ` + `TRUNCATE_EXISTING` behave on Windows.

Reply:
WASI says this currently:

> If __WASI_RIGHT_PATH_OPEN is set, includes the right to invoke __wasi_path_open() with __WASI_O_TRUNC.

So this change was in accordance to that.

I think that requiring the right to write to the file would make sense since all the underlying operating systems do so.

@sunfishcode your thoughts?

Reply:
Oh and on Windows, `CreateFile` requires the literal passing of `GENERIC_WRITE` when opening for truncation, otherwise it errors with invalid argument.

Reply:
I should also mention the expectation that `O_TRUNC` works *without* `__WASI_RIGHT_FD_WRITE` is the reason #570 is failing, since that's what `truncation_rights` is doing.

Reply:
> WASI says this currently:
> 
> > If __WASI_RIGHT_PATH_OPEN is set, includes the right to invoke __wasi_path_open() with __WASI_O_TRUNC.
> 
> So this change was in accordance to that.
> 
> I think that requiring the right to write to the file would make sense since all the underlying operating systems do so.
> 

But this is the spec for `__WASI_RIGHT_PATH_FILESTAT_SET_SIZE`. But this means that we probably should be more careful in #570.

Reply:
Ah I think I misread that paragraph then.  It isn't `__WASI_RIGHT_PATH_OPEN ` that is inferring the right for `__WASI_O_TRUNC`, but the combination of the two.

In that case, I will back out this change and fix the test case (unless you beat me to it in #570).

Reply:
So I removed this workaround to the test failure, so the `truncation_rights` test will now fail this CI.

I'm waiting on clarification to questions posed in #570 to see how we should be fixing this.

Reply:
I pushed up a reversion for disabling the `truncation_rights` for Windows again.  I think we should reenable it as part of #570.

Reply:
I pushed up a reversion for disabling the truncation_rights for Windows again. I think we should reenable it as part of #570.